src/Compilers/CSharp/Portable/Symbols/Symbol.cs

```diff
@@ -971,31 +971,40 @@ internal bool DeriveUseSiteDiagnosticFromType(ref DiagnosticInfo result, TypeSym
             {
                 if (info.Code == (int)ErrorCode.ERR_BogusType)
                 {
-                    switch (this.Kind)
-                    {
-                        case SymbolKind.Field:
-                        case SymbolKind.Method:
-                        case SymbolKind.Property:
-                        case SymbolKind.Event:
-                            info = new CSDiagnosticInfo(ErrorCode.ERR_BindToBogus, this);
-                            break;
-                    }
+                    info = GetSymbolSpecificUnsupprtedMetadataUseSiteErrorInfo() ?? info;
                 }
             }
 
             return MergeUseSiteDiagnostics(ref result, info);
         }
 
-        internal bool DeriveUseSiteDiagnosticFromType(ref DiagnosticInfo result, TypeWithAnnotations type)
+        private DiagnosticInfo GetSymbolSpecificUnsupprtedMetadataUseSiteErrorInfo()
+        {
+            switch (this.Kind)
+            {
+                case SymbolKind.Field:
+                case SymbolKind.Method:
+                case SymbolKind.Property:
+                case SymbolKind.Event:
+                    return new CSDiagnosticInfo(ErrorCode.ERR_BindToBogus, this);
+            }
+
+            return null;
+        }
+
+        internal bool DeriveUseSiteDiagnosticFromType(ref DiagnosticInfo result, TypeWithAnnotations type, AllowedRequiredModifierType allowedRequiredModifierType)
         {
             return DeriveUseSiteDiagnosticFromType(ref result, type.Type) ||
-                   DeriveUseSiteDiagnosticFromCustomModifiers(ref result, type.CustomModifiers);
+                   DeriveUseSiteDiagnosticFromCustomModifiers(ref result, type.CustomModifiers, allowedRequiredModifierType);
         }
 
         internal bool DeriveUseSiteDiagnosticFromParameter(ref DiagnosticInfo result, ParameterSymbol param)
         {
-            return DeriveUseSiteDiagnosticFromType(ref result, param.TypeWithAnnotations) ||
-                   DeriveUseSiteDiagnosticFromCustomModifiers(ref result, param.RefCustomModifiers);
+            return DeriveUseSiteDiagnosticFromType(ref result, param.TypeWithAnnotations, AllowedRequiredModifierType.None) ||
+                   DeriveUseSiteDiagnosticFromCustomModifiers(ref result, param.RefCustomModifiers,
+                                                              this is MethodSymbol method && method.MethodKind == MethodKind.FunctionPointerSignature ?
```

Comment:
It feels like this would be better as a parameter to `DeriveUseSiteDiagnosticFromParameter`, rather than calculated internally here.

Reply:
> It feels like this would be better as a parameter to DeriveUseSiteDiagnosticFromParameter, rather than calculated internally here.

I do not think so. We have all information to reliably make the decision here, why would we want the consumer to to do extra work, given that the condition is still going to be pretty much the same.

---
In reply to: [441015421](https://github.com/dotnet/roslyn/pull/45220#discussion_r441015421) [](ancestors = 441015421)

Reply:
> We have all information to reliably make the decision here, why would we want the consumer to to do extra work, given that the condition is still going to be pretty much the same.

Because it feels like this is the wrong place to make that decision. We can enumerate the possibilities here today, but that feels fragile. It seems like the best place to state the required modifiers is the component that actually cares about that requirement. I'm not going to block this PR on it as it's minor, but it really doesn't feel like the appropriate location.

---
In reply to: [441020808](https://github.com/dotnet/roslyn/pull/45220#discussion_r441020808) [](ancestors = 441020808,441015421)

Reply:
> Because it feels like this is the wrong place to make that decision. We can enumerate the possibilities here today, but that feels fragile. It seems like the best place to state the required modifiers is the component that actually cares about that requirement. I'm not going to block this PR on it as it's minor, but it really doesn't feel like the appropriate location.

Even if we assume that the code is fragile, which I do not agree with. I am failing to see how duplication of the same logic in many places makes the code less fragile.


---
In reply to: [441029178](https://github.com/dotnet/roslyn/pull/45220#discussion_r441029178) [](ancestors = 441029178,441020808,441015421)